### PR TITLE
fix(binder): fix function call binding

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Previously our `bind_function` is a mess. An example is #17560, which is because we bind UD(A/T)F and window function in different path, and we never considered a combination of the two.

This PR re-organizes `bind_function` and is a pre-requisition for fixing #17560.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by stdrc** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18177?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#18177** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18177?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#18130** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18130?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @stdrc and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/expr/core/src/window_function/kind.rs
Comment (kwannoel) [inline src/expr/core/src/window_function/kind.rs:22]: What is "for builtin"

File: src/frontend/planner_test/tests/testdata/output/over_window_function.yaml
Comment (kwannoel) [inline src/frontend/planner_test/tests/testdata/output/over_window_function.yaml:65]: Seems this error message is less precise?

File: src/expr/core/src/window_function/kind.rs
Comment (stdrc) [reply src/expr/core/src/window_function/kind.rs:22]: We can only parse builtin window functions directly from a string.

File: src/frontend/src/binder/expr/function/mod.rs
Comment (kwannoel) [inline src/frontend/src/binder/expr/function/mod.rs:129]: Seems heavy to me to introduce a macro instead of just a function to construct 
```
            return Err(ErrorCode::InvalidInputSyntax($msg.to_string()).into());
```

But I'm not going to block this PR with it.

**Review by kwannoel — APPROVED** [review]
LGTM, thanks!

File: src/frontend/planner_test/tests/testdata/output/over_window_function.yaml
Comment (stdrc) [reply src/frontend/planner_test/tests/testdata/output/over_window_function.yaml:65]: To me, the two error messages both look good. Actually I think it's equally possible that the user 1) forgets to write `OVER` clause when intended to do a window function call or 2) mistypes the normal scalar function `log` as `lag`.

Comment (stdrc) [reply src/frontend/planner_test/tests/testdata/output/over_window_function.yaml:65]: May need additional work to achieve msg like `function lag(integer) does not exist as a scalar function, do you mean log or do you forget the OVER clause` which is better. We can leave it to future.

File: src/frontend/src/binder/expr/function/mod.rs
Comment (BugenZhao) [inline src/frontend/src/binder/expr/function/mod.rs:1]: The diff of this file on GitHub is a mess as well. 🤣

**Pull request merged**
